luda-adventure/src/app/game/types/game_object/types/positioner/movement_plan.rs

```diff
@@ -0,0 +1,59 @@
+use super::{CollisionState, Movement};
+use crate::app::game::Tile;
+use namui::prelude::*;
+
+pub struct MovementPlan {
+    pub original_movement: Movement,
+    pub predicted_movement_list: Vec<Movement>,
```

Comment:
I suggest making sure about this data.


There are two type of plan,
- The plan the user or somebody want to move or expected
- The real movement plan which will be happen in future absolutely

And we strongly assumed that no item will be moved, but only character will be moved.

Then there is only two types of movement result,
1. fixed position (when it is not character or player didn't move)
2. moving (only for character when player lets character move)

So I believe we can make it simpler like this.
Think about it and let me your opinion.

```rust
enum Motion {
    Fixed(Xy),
    Moving(Vec<MovePath>)
}

struct MovePath {
    start_xy: Xy,
    start_time: Time,
    end_xy: Xy,
    end_time: Time,
}

struct Positioner {
    motion: Motion
}

impl Positioner {
    pub fn xy(&self, time: Time) -> Xy {
        match self.motion {
            Motion::Fixed(xy) => xy,
            Motion::Moving(paths) => {
                paths.iter().find(|path| {
                    path.start_time <= time && time <= path.end_time
                }).map(|path| {
                    let ratio = (time - path.start_time) / (path.end_time - path.start_time);
                    Xy {
                        x: path.start_xy.x + (path.end_xy.x - path.start_xy.x) * ratio,
                        y: path.start_xy.y + (path.end_xy.y - path.start_xy.y) * ratio,
                    }
                }).unwrap_else(|| unreachable!())
            }
        }
    }
    pub fn update_motion(&mut self, motion: Motion) {
        self.motion = motion;
    }
}

impl App {
    fn update(&mut self, event: &dyn std::any::Any) {
        if is_time_to_update_character_motion { // when user change input axis or it needs to update prediction by time spending
            let (character, character_positioner) = self.ecs_app.find_character_and_its_positioner();
            let next_motion = character.calculate_next_motion(&self.ecs_app);
            character_positioner.update_motion(next_motion);
        }
    }
}

[Component]
struct Character {
    user_input_axis: Xy,
}

impl Character {
    pub fn calculate_next_motion(&self, ecs_app: &EcsApp) -> Motion {
        // get all objects
        // and calculate them at here all.
    }
}
```

Reply:
Your suggestion looks more appropriate for ecs and our design.

Reply:
`calculate_next_motion` requires not only character components but also positioners and colliders. Therefore, I think it is appropriate for `calculate_next_motion` to be implemented outside rather than on character components. I think this is similar to ECS's system.


```rust
enum Motion {
    Fixed(Xy),
    Moving(Vec<MovePath>)
}
```
But I like this concept, and I think I need a diet in my current logic, so I'm trying to apply your suggestion.

Reply:
What's the purpose to put that function outside of character component? I already thought about it so I put `ecs_app: &EcsApp` as parameter to query all of them.

Reply:
For simplicity.
If we implement a function in a character component, We need to query the entity that owns this component and re-query the components that it has. Can we say this is a function of the character component?

I think it should be a system that uses character, positioner, colider components.
![image](https://user-images.githubusercontent.com/38313680/195758594-739c06cb-6416-4c89-8ab8-cc4e4c029f0f.png)

Reply:
We right now have no system.

It's not a big deal to query inside of character component's function.

It would be better to put function inside of character component because it's the only place that function will be used and gathering together by interest is good practice in my thought.

Reply:
Good. It's not a big deal.  I'll focus on `gathering together by interest` and continue.

Reply:
Hm.... I think it's ok to seperate all function from component and put them into one directory like a `systems`. Because they have not only single component's system but also combination's one. I'm not sure about it. go a head what you want.

Reply:
It acts like a `system`, but the name will be use `update`. Because it is more suitable for namui.